### PR TITLE
Fix filter polling leak

### LINKED ISSUE #1458: Filter Subprovider Does Not Stop Polling When Page Is Closed
When a site creates a `filter`, since Infura is stateless, we polyfill it client-side with [filter subprovider](https://github.com/MetaMask/provider-engine/blob/master/subproviders/filters.js), and poll for new logs on each new block.

However, when a site is closed, we should stop polling for those logs. Currently we seem to continue polling indefinitely.

This is especially noticeable when using the ENS registrar dapp, which adds a filter for each domain you visit and check out.

This may be a regression, or this may be a very old bug (it must be a regression, right???)

### PR DESCRIPTION
**Pull request by danfinlay**
By injecting a `FilterSubprovider` into the `inpage-provider`, I'm able to keep filter polling logic in-page, while keeping caching logic in our background process, so that subscriptions are deallocated when pages are closed.

Fixes #1458

Required refactoring the `StreamingSubprovider` to be a composable `ProviderEngine` instance, so that the `FilterSubprovider` could be added to it.

For this reason, there was a small API break in `web3-stream-provider` that needs to be merged & deployed before these tests can be rebuilt & will pass:
https://github.com/kumavis/web3-stream-provider/pull/1

### CODE REVIEW

File: app/scripts/lib/inpage-provider.js [hunk 1]
```diff
@@ -27,14 +29,24 @@ function MetamaskInpageProvider (connectionStream) {
   )
 
   // connect to async provider
-  const asyncProvider = self.asyncProvider = new StreamProvider()
+  const engine = new ProviderEngine()
+
+  const filterSubprovider = new FilterSubprovider()
+  engine.addProvider(filterSubprovider)
+
+  const stream = self.stream = new StreamSubprovider()
```
Comment (kumavis) [inline app/scripts/lib/inpage-provider.js]: can you rename to `streamProvider` or something less generic

Comment (flyswatter) [inline app/scripts/lib/inpage-provider.js]: Like `RemoteStreamSubprovider`?  I'm not sure what you mean less generic. Isn't descriptive good here?

Comment (kumavis) [inline app/scripts/lib/inpage-provider.js]: i mean instead of `stream`

Comment (flyswatter) [inline app/scripts/lib/inpage-provider.js]: Oh ok, done.

**Comment by flyswatter** [comment]
Good changes! 👍

**Review by kumavis — APPROVED** [review]

**Pull request merged**